tools/clang/tools/dxv/dxv.cpp

```diff
@@ -80,19 +106,50 @@ void DxvContext::Validate() {
       std::string msg(pStart);
       IFTMSG(status, msg);
     } else {
+      // Source was unsigned DxilContainer, write signed container if it's now signed:
+      if (bSourceIsDxilContainer && bSourceIsUnsigned) {
+        const hlsl::DxilContainerHeader *pHeader = hlsl::IsDxilContainerLike(
+            pSource->GetBufferPointer(), pSource->GetBufferSize());
+        hlsl::DxilContainerHash blankHash = {};
+        if (memcmp(&pHeader->Hash, &blankHash,
+                   sizeof(hlsl::DxilContainerHash)) != 0) {
+          if (OutputFilename.empty()) {
+            OutputFilename.assign(InputFilename);
+          }
+          WriteBlobToFile(pSource, StringRefUtf16(OutputFilename), CP_ACP);
```

Comment:
I don't see any check that the validation is from dxil.dll and not the internal validator. Did I miss it?

Reply:
We check that it was unsigned before, and now it's signed, which only happens with DXIL.dll.  That's really all we can know.  We will be arriving at DXIL.dll through DxCompiler.dll here.  That means it's up to DxCompiler.dll to pass things through to DXIL.dll, and there's basically no other indication that we used DXIL.dll from this code.  This also means if you run dxv.exe without dxcompiler.dll present, it won't work, even though it will only call through to DXIL.dll.  It uses DxCompiler.dll for assembler as well.  So if we want it to work with only DXIL.dll, we would have to modify the code somewhat to attempt to initialize from DxCompiler.dll, then failing that set a flag and initialize from DXIL.dll.  If the flag is set, we would only support the full container validation, since we wouldn't have the assembler and can't use DXIL.dll for bitcode validation.